hca/dss/__init__.py

```diff
@@ -107,87 +109,182 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
+
+                        consume_bytes = int(fh.tell())
+                        server_start = 0
+                        content_range_header = response.headers.get('Content-Range', None)
+                        if content_range_header is not None:
+                            cre = re.compile("bytes (\d+)-(\d+)")
+                            mo = cre.search(content_range_header)
+                            if mo is not None:
+                                server_start = int(mo.group(1))
+
+                        consume_bytes -= server_start
+                        assert consume_bytes >= 0
+                        if server_start > 0 and consume_bytes == 0:
+                            logger.info("%s", "File {}: Resuming at {}.".format(
+                                file_uuid, server_start))
+                        elif consume_bytes > 0:
+                            logger.info("%s", "File {}: Resuming at {}. Dropping {} bytes to match".format(
+                                file_uuid, server_start, consume_bytes))
+
+                            while consume_bytes > 0:
+                                bytes_to_read = min(consume_bytes, 1024*1024)
+                                content = response.iter_content(chunk_size=bytes_to_read)
+                                chunk = next(content)
+                                if chunk:
+                                    consume_bytes -= len(chunk)
+
+                        for chunk in response.iter_content(chunk_size=1024*1024):
+                            if chunk:
+                                fh.write(chunk)
+                                hasher.update(chunk)
+                                retries_left = min(retries_left + 1, num_retries)
+                                delay = max(delay / 2, min_delay_seconds)
+                        break
+                    finally:
+                        response.close()
+                except (ChunkedEncodingError, ConnectionError, ReadTimeout):
+                    if retries_left > 0:
+                        logger.info("%s", "File {}: GET FAILED. Attempting to resume.".format(
+                            file_uuid, dest_path))
+                        time.sleep(delay)
+                        delay *= 2
+                        retries_left -= 1
+                        continue
+                    raise
+
+        if hasher.hexdigest().lower() != file_sha256.lower():
+            os.remove(dest_path)
+            logger.error("%s", "File {}: GET FAILED. Checksum mismatch.".format(file_uuid))
+            raise ValueError("Expected sha256 {} Received sha256 {}".format(
+                file_sha256.lower(), hasher.hexdigest().lower()))
+        else:
+            logger.info("%s", "File {}: GET SUCCEEDED. Stored at {}.".format(file_uuid, dest_path))
+
+    @classmethod
+    def _file_path(cls, checksum):
+        """
+        returns a file's relative local path based on the nesting parameters and the files hash
```

Comment:
Comment could be more descriptive.  As in, given a string: `64f0b864f0b8e7fc22a8c5f7ec6649530a03e4` it will return `64/f0b8/64f0b8e7fc22a8c5f7ec6649530a03e4` to be used as a directory structure for the file, with the slashes always where `FILE_NESTING_PARAMETER` indicates.

Also, will the user be getting a ton of obscurely named (from there point of view) directories?  If so, should we rethink this?

Also, why create multiple directories?  Why not just one directory with the checksum name, as in `FILE_NESTING_PARAMETER=[len(checksum)]`?

Reply:
Is the epic #232?  I'm having trouble seeing much information there.  I'd still like to argue for simplifying this by not creating the extra dirs if that's not too unreasonable.  I really like simple code when one can get away with it.

Reply:
Yeah, the epic doesn't make the plan clear, sorry. I talked about this at the F2F, though, and it was discussed at length on Slack in `#cli-download-layout`. Feel free to talk to me personally, if you have more questions about the big picture. 

The point is that we want to maintain a content-addressed file store that could hold hundreds of thousands of files. We can't put them all in the same directory because that would mean even running `ls` will take minutes to complete.